### ISSUE TITLE
Controls overlap in video quick settings panel

### ISSUE THREAD

**Issue by low-batt** [iina/iina#3790]
**System and IINA version:**

- macOS 12.4
- IINA 1.3.0

**Expected behavior:**
UI controls do not overlap.

**Actual behavior:**
In the video quick settings panel controls under "Aspect Ratio" and "Crop" overlap and are truncated:

![video-quick-settings](https://user-images.githubusercontent.com/86170219/172025365-d48b5687-08bb-4c8c-ab3c-3e7eef4428e8.jpg)

Because of that problem I did not recognize the blank blob at the end of the list of aspect ratios as a text field that can be used to enter a custom ratio. Due to the overlapping I thought it was some sort of broken button. One way to make it obvious what this grey blob is would be to populate that field with the current value like is done in the "Speed" setting. If the text field is moved to a following line to correct the overlap, then a "Custom:" label can be added before it.

Note that when run under Xcode the `QuickSettingView` triggers constraint errors:
```text
2022-06-04 16:46:21.727556-0400 IINA[49559:2719443] [Layout] Unable to simultaneously satisfy constraints:
(
    "<NSLayoutConstraint:0x6000013ed8b0 H:[NSView:0x15c828280]-(0)-|   (active, names: '|':IINA.FlippedView:0x15c828040 )>",
    "<NSLayoutConstraint:0x6000013ed720 H:|-(0)-[NSView:0x15c828280]   (active, names: '|':IINA.FlippedView:0x15c828040 )>",
    "<NSLayoutConstraint:0x6000013ed6d0 H:[IINA.FlippedView:0x15c828040]-(0)-|   (active, names: '|':NSClipView:0x15c82b4e0 )>",
    "<NSLayoutConstraint:0x6000013ed630 H:|-(0)-[IINA.FlippedView:0x15c828040]   (active, names: '|':NSClipView:0x15c82b4e0 )>",
    "<NSAutoresizingMaskLayoutConstraint:0x6000013f8e60 h=-&- v=-&- NSClipView:0x15c82b4e0.minX == 0   (active, names: '|':NSScrollView:0x15c0ed000 )>",
    "<NSAutoresizingMaskLayoutConstraint:0x6000013f9220 h=-&- v=-&- NSView:0x15c82af70.minX == 0   (active, names: '|':NSTabView:0x15c82a890 )>",
    "<NSAutoresizingMaskLayoutConstraint:0x6000013f9270 h=-&- v=-&- H:[NSView:0x15c82af70]-(0)-|   (active, names: '|':NSTabView:0x15c82a890 )>",
    "<NSLayoutConstraint:0x6000013ed680 H:|-(0)-[NSScrollView:0x15c0ed000]   (active, names: '|':NSView:0x15c82af70 )>",
    "<NSLayoutConstraint:0x6000013ed4f0 H:[NSScrollView:0x15c0ed000]-(0)-|   (active, names: '|':NSView:0x15c82af70 )>",
    "<NSAutoresizingMaskLayoutConstraint:0x6000013f8eb0 h=-&- v=-&- H:[NSClipView:0x15c82b4e0]-(15)-|   (active, names: '|':NSScrollView:0x15c0ed000 )>",
    "<NSLayoutConstraint:0x6000013e57c0 H:|-(21)-[NSSlider:0x12af5bf60]   (active, names: '|':NSView:0x15c828280 )>",
    "<NSLayoutConstraint:0x6000013e5db0 H:[NSTextField:0x12af50cf0]-(24)-|   (active, names: '|':NSView:0x15c828280 )>",
    "<NSLayoutConstraint:0x600001317e80 NSSlider:0x12af5bf60.width == 236   (active)>",
    "<NSLayoutConstraint:0x6000013e5bd0 H:[NSSlider:0x12af5bf60]-(10)-[NSTextField:0x12af509b0]   (active)>",
    "<NSLayoutConstraint:0x600001317bb0 NSTextField:0x12af509b0.width >= 57   (active)>",
    "<NSLayoutConstraint:0x6000013e5d60 H:[NSTextField:0x12af509b0]-(5)-[NSTextField:0x12af50cf0]   (active)>",
    "<NSLayoutConstraint:0x6000013fd270 H:[NSTabView:0x15c82a890]-(0)-|   (active, names: '|':IINA.QuickSettingView:0x15c829630 )>",
    "<NSLayoutConstraint:0x6000013fd360 H:|-(0)-[NSTabView:0x15c82a890]   (active, names: '|':IINA.QuickSettingView:0x15c829630 )>",
    "<NSLayoutConstraint:0x6000013fd040 IINA.QuickSettingView:0x15c829630.width == 360   (active)>"
)

Will attempt to recover by breaking constraint 
<NSLayoutConstraint:0x600001317bb0 NSTextField:0x12af509b0.width >= 57   (active)>

Set the NSUserDefault NSConstraintBasedLayoutVisualizeMutuallyExclusiveConstraints to YES to have -[NSWindow visualizeConstraints:] automatically called when this happens.  And/or, set a symbolic breakpoint on LAYOUT_CONSTRAINTS_NOT_SATISFIABLE to catch this in the debugger.
```

**Steps to reproduce:**
Open the video quick settings panel.

- [x] MPV does not have this problem.

**How often does this happen?**
Everytime.

**Comment by svobs** [comment]
Although this is maybe the most glaring problem, this panel needs a lot of work. There are a lot of XCode warnings about possible clipping throughout the panel due to fixed or missing layout constraints. It looks like almost none (or maybe none?) of the elements in this panel have constraints which relate the spacing or alignment to their neighbors.

Even though MacOS no longer allows users to change OS fonts, and the elements in the panel have been laid out as fixed down to the pixel, some elements may still need to move slightly and need guidance for how to do that. XCode points out that different localizations may render the labels differently in different localizations and these will inevitably have different sizes. For example, "Brightness" might be an even longer word in some other language, and when that happens we'd probably want the width of all the sliders to be reduced by a few pixels, taking from the left side, but currently it would just spill over onto the slider. And even if all the controls look fine in all languages, Apple may change their sizes slightly in new MacOS releases which may cause clipping or misalignment. This already seems to have happened with some of the other views: XCode has some warnings about widths and heights being a few pixels different than expected, but in most cases it's using the constraints to correct them when they are laid out so the end result looks ok.

That said, I'm not seeing this particular clipping in 1.3.0 Build 132 / MacOS 12.4... I wonder what could be causing the difference?
<img width="358" alt="Screen Shot 2022-06-04 at 15 46 25" src="https://user-images.githubusercontent.com/2213815/172027742-dfd7841b-856b-4d82-8840-d073262b5b8d.png">

**Comment by low-batt** [comment]
I was wondering why this had not been reported before. I'm using macOS 12.4 on one of the new 16" MacBook Pros with the M1 chip. I was seeing this before 1.3.0, so not a new regression. I also see this on my Catalina machine. Maybe some global setting I usually change? I'll look into that.

**Comment by low-batt** [comment]
I poked around and I'm left confused as to why it is working for you and messing up for me.

Anyone else out there reading this issue please report your experience, especially if you see the problem I'm seeing.

**Comment by low-batt** [comment]
Found the configuration that causes it to fail for me. I found scroll bars disappearing and appearing distracting, so I configure this in System Preferences/General:

![scroll-bars-preference](https://user-images.githubusercontent.com/86170219/172030414-69aff5a7-d71e-405e-a6d9-854ae4acabdb.jpg)

**Comment by svobs** [comment]
@low-batt Ah, interesting! I was able to reproduce the problem. Yeah, I leave them hidden and it looks nice, but in the 5% of the time I need to grab them, they are hard to grab ahold of, so none of the options are perfect.

Well, I played with the .xib file some, and here are some things I came up with.

Tried breaking it into 2 lines but don't think it looks too great - there's too much unused space.
![Screen Shot 2022-06-04 at 22 26 56](https://user-images.githubusercontent.com/2213815/172038432-3f60ea78-d192-421b-a8ce-b9eb4ee0ae69.png)
_Not an actual runtime screenshot - just a screenshot from Interface Builder, sorry._

My vote for the way to go would be to add a placeholder, which is shown in gray when there is nothing entered. I changed the style of the text field and the segmented control next to it to rectangular so they would align perfectly, although it does look a bit utilitarian, and the other controls would need to be changed to match.
![Screen Shot 2022-06-04 at 22 58 12](https://user-images.githubusercontent.com/2213815/172038534-b9a580dd-acf7-4249-beb8-3d781da41eab.png)
_With placeholder. Needs vertical alignment fix, but this is just an idea_

![Screen Shot 2022-06-04 at 22 59 11](https://user-images.githubusercontent.com/2213815/172038597-b077e5ac-81cd-4cf3-a6bc-d71b98b37d50.png)
_With entry_

Another idea is to hide the borders entirely:
![Screen Shot 2022-06-04 at 23 23 09](https://user-images.githubusercontent.com/2213815/172038666-28bc3136-ec1c-4fc4-837a-10f7c94beec9.png)
_No border. Showing placeholder_

![Screen Shot 2022-06-04 at 23 24 21](https://user-images.githubusercontent.com/2213815/172038669-2a731e21-4523-4cb6-b01c-add4859815ca.png)
_No border. Focus ring = default_

There should be a way to use styling so that a custom selection looks like a selected button.

**Comment by svobs** [comment]
Upon revisiting this, I think that was all over-design 😂  The same design but with alignment & constraint fixes and placeholder text are probably the way to go.

@low-batt let me know if you'd like to fix this, or I can try to do it since I still have a lot of free time. My concern is that XIB files are not easily mergable, so am hesitant to put a lot of effort into one unless it has a high likelihood of getting merged into develop relatively quickly and before other changes need to be made to the file.

**Comment by low-batt** [comment]
I'll let you handle this one. The timing of making this fix definitely needs to be coordinated with @lhc70000 to avoid the issues with XIB merge conflicts you mentioned.

So for the moment avoid coding this up.

**Comment by low-batt** [comment]
This is a violation of Apple's [Human Interface Guidelines](https://developer.apple.com/design/human-interface-guidelines/guidelines/overview). From section [scroll-views](https://developer.apple.com/design/human-interface-guidelines/components/presentation/scroll-views):
> ## macOS
> 
> **Account for scroll bars in your layout.** By default, scroll bars appear only when people interact with views that contain them, but people can use a setting in General settings to make them to appear all the time. Some input devices also cause scroll bars to display all the time. If necessary, adjust the layout of your window so important interface elements don't appear beneath scroll bars. The scroll bar track has a thickness of 15 points (regular size) or 11 points (small or mini size).

This definitely needs to be fixed.

**Comment by svobs** [comment]
> This definitely needs to be fixed.

Agreed.  I actually started working on it a couple of days ago. I suppose it's possible to come up with a quick fix like adding a little more fixed horizontal space, or adding a new hidden view which expands or something like that, but those aren't very durable solutions. I wanted to replace all the constraints (at least in the Video Settings panel) with ones which knows how to adapt to changes in components size generally.    

I expected to post something here earlier, but it's been even more of a slog than I expected. I thought I had gotten decent at Interface Builder but I've been getting schooled a lot more 😅  .... I think a big part of the steep learning curve is the fact that Auto Layout was created long after most of  the components, and it's not a perfect fit. How each component type responds depends on its own internal rules and limitations, and that results in ton of special cases. Worse, when faced with ambiguity the layout engine will just pick the wrong layout (or crash XCode, often enough).

Made some good progress yesterday though. There are still a few things I need to dig into. The switches are using some custom code, and there are a bunch of constraints which are being manipulated programmatically, so I need to examine those carefully. But I should have something in the next few days.

**Comment by svobs** [comment]
Just noticed this. Does this look correct?

![lines](https://user-images.githubusercontent.com/2213815/197420241-1be6e870-7585-4910-b6cb-60f20152c5f7.png)

Seems like the horizontal separators should go all the way from the left to the right of the box, no? So the left should look like the right? Or maybe the right should look like the left. But not the way it is now

**Comment by low-batt** [comment]
That definitely is wrong in my opinion.

Looking at divider lines in other windows, such as the preference window, the convention seems to be to stop the line at the end of contents and not have the line span the entire width of the window.

So I'm thinking the left is correct and the right should stop short and line up with the end of the buttons.

Although this issue has increased priority now what we know it is also a violation of the HIG, at the moment the review of PR #3887 is of higher priority. The developers are actively reviewing that one and looking to get that fix merged. Please have a look at their comments and respond.

**Comment by svobs** [comment]
Update on this... it's been a ton of effort, but I've redone the layout for the whole panel so that things shrink and grow properly, and got rid all the warnings... but it's still not 100% with respect to the scroll bars. Apple did not make this easy.

Take a look at [my branch](https://github.com/svobs/iina/tree/pr-video-settings-layout) if you want to try it yourself, but here's a visual summary.

Here it is when `Show scroll bars` is set to `When scrolling`. I copied and pasted little red ticks to make it a little easier to see. Sorry I didn't do the best job of cropping - you'll have to believe me that the sides end where they do.

![Screen Shot 2022-10-25 at 17 21 24](https://user-images.githubusercontent.com/2213815/197908034-38b9bd61-de6e-473e-bee6-38562008b93e.png)
As you can see, it's nearly perfect, but even though all the nested frames are constrained to line up exactly, there are still 2 extra pixels which allow a tiny bit of horizontal scroll. I suspect one pixel might be getting added to each side by the `NSClipView` and subtracting those from its visible content view when it does the overlay.

When `Show scroll bars` is set to `Always`, the problem is worse because the system adds additional horizontal space for the scroll bars, although it's not immediately obvious:
![Screen Shot 2022-10-25 at 17 22 20](https://user-images.githubusercontent.com/2213815/197908896-bc869af3-9475-4f22-ba0a-0be187925f48.png)

But if I scroll left or right it becomes more obvious:
![Screen Shot 2022-10-25 at 17 22 32](https://user-images.githubusercontent.com/2213815/197910029-2a24d48e-ebdc-4576-92c0-16b8a6ea6a57.png)

This is better than before. But what I was hoping to do was shrink each of the components horizontally, not add horizontal scrolling...

In Interface Builder, a scroll view is represented by two components, `NSScrollView` and an `NSClipView`. The dimensions of the former represents the outer dimensions which includes the scroller(s), and the latter's dimensions represent the full size of the content which goes inside it, which can be a lot larger. But neither of those give us the viewport / the visible content dimensions which don't include the scrollers. I think I need to set a constraint programmatically to do that, so that will be my next focus.

Also - I tested all the functionality pretty thoroughly to make sure I didn't break anything, but there's some odd plugin-related layout which I tried not to touch. It looks like there are 2 invisible tabs for plugins. I'm not sure if both or either are being used. If anyone has an example of a plugin which adds something to the side panel, that would be really helpful.

**Comment by svobs** [comment]
Another update. I just noticed that Audio and Equalizer pages are throwing out constraint errors, so I think I'll end up cleaning them up too.

> Note that when run under Xcode the QuickSettingView triggers constraint errors:

I found you can actually see these from the last 1.3.0 release - they aren't picked up in the IINA logs but you can see them from the system logs. Do the following:

1. Open Console.app.
2. Enter "iina" in the Search box and hit `Return` to activate the filter.
3. Click the Start icon to start the stream.
4. Go into IINA, open a video, and open the Quick Settings panel. Click to change between Video, Audio, Playlist tabs. You should see it lines start to appear in Console.

<img width="1165" alt="Screen Shot 2022-10-26 at 13 47 37" src="https://user-images.githubusercontent.com/2213815/198135383-69633804-3f13-4d60-8853-8060b3676345.png">

**Comment by svobs** [comment]
I've been attacking this pretty hard, and ended up rewriting the layout 3 times over, but I've finally gotten the Video Settings panel just about right. So, 2 more to go. Hopefully they won't be as bad...

I had to make a hard call to make it work though. The problem turned out to be a lack of horizontal space. I was able to nip & tuck things to make it work - except for the **Crop** section.

![Screen Shot 2022-10-27 at 20 01 51](https://user-images.githubusercontent.com/2213815/198500116-f1c28cc0-6c72-4fb5-bce8-df841c14fecd.png)

My solution was to change the button from the verbose `Custom...` to a more nebulous `...` so it would fit, but I'm not in love with it. Some other possibilities I was thinking about:
1.  Reduce the font size of the button (tried this; it's not enough by itself) `and` in the segmented controls, freeing up the ~8 pixels needed.
2. Make the Crop section multi-line somehow.
3. Change "None" into something smaller?
4. Resize the panel¹ ... perhaps another time.

¹ In theory, expanding the Quick Settings panel to 380 pixels from its current 360 would be a lot easier to work with, and would still be reasonable given the Playlist/Chapters panel's current range of 240 to 400 pixels. So I actually attempted this... Unfortunately, I ended up scrambling the layout so badly that I couldn't make it sane again. I ended up scrapping the whole thing and checking out a fresh copy. (You really learn to lean heavily on Git when you're using Interface Builder...) I've maybe learned a lot more about IB since then, but I just can't back down that path right now.

**Comment by svobs** [comment]
Actually how about this? This was a trivial change.

![Screen Shot 2022-10-27 at 22 04 41](https://user-images.githubusercontent.com/2213815/198507867-ef57ba49-fb45-472a-ae72-35e4058e33f1.png)
After selecting it, and opening up the panel again:
![Screen Shot 2022-10-27 at 22 07 30](https://user-images.githubusercontent.com/2213815/198507871-0ed00ff5-cd8a-4636-93a8-29bff37e9adb.png)

**Comment by saagarjha** [comment]
How does this segmented control work? The button lets you click it at any time…

**Comment by svobs** [comment]
> How does this segmented control work? The button lets you click it at any time…

Instead of "Custom..." having a separate button, I made it a button in the segmented control, which is really just a group of buttons anyway, and then I also made it show that button as selected/depressed when a custom crop is active. I felt like it was an improvement over the current behavior (which is to just deselect all the buttons when a custom crop is detected) because it indicates a status. But clicking on the button will still do what it did before (interactive mode?) even if it's already selected.
